openvdb_cmd/vdb_tool/include/Parser.h

```diff
@@ -0,0 +1,1041 @@
+    virtual bool next() = 0;
+    template <typename T>
+    T get() const { return strTo<T>(memory.get(name)); }
+    template <typename T>
+    void set(T v){
+        memory.set(name, v);
+        memory.set("#"+name, pos);
+    }
+    void print(std::ostream& os = std::cerr) const {
+        os << "Processing: " << name << " = " << memory.get(name) << ", counter #" << name << " = " << pos <<std::endl;
+    }
+};// BaseLoop struct
+
+// ==============================================================================================================
+
+template <typename T>
+struct ForLoop : public BaseLoop
+{
+    using BaseLoop::pos;
+    math::Vec3<T> vec;
+public:
+    ForLoop(Memory &s, ActIterT i, const std::string &n, const std::vector<T> &v) : BaseLoop(s, i, n), vec(1) {
+        if (v.size()!=2 && v.size()!=3)  throw std::invalid_argument("ForLoop: expected two or three arguments, i=1,9 or i=1,9,2");
+        for (size_t i=0; i<v.size(); ++i) vec[i] = v[i];
+        if (this->valid()) this->set(vec[0]);
+    }
+    virtual ~ForLoop() {}
+    bool valid() override {return vec[0] < vec[1];}
+    bool next() override {
+        ++pos;
+        vec[0] = this->template get<T>() + vec[2];// read from memory
+        if (vec[0] < vec[1]) this->set(vec[0]);
+        return vec[0] < vec[1];
+    }
+};// ForLoop struct
+
+// ==============================================================================================================
+
+class EachLoop : public BaseLoop
+{
+    using BaseLoop::pos;
+    const VecS vec;// list of all string values
+public:
+    EachLoop(Memory &s, ActIterT i, const std::string &n, const VecS &v) : BaseLoop(s,i,n), vec(v.begin(), v.end()){
+        if (this->valid()) this->set(vec[0]);
+    }
+    virtual ~EachLoop() {}
+    bool valid() override {return pos < vec.size();}
+    bool next() override {
+        if (++pos < vec.size()) this->set(vec[pos]);
+        return pos < vec.size();
+    }
+};// EachLoop class
+
+// ==============================================================================================================
+
+class IfLoop : public BaseLoop
+{
+public:
+    IfLoop(Memory &s, ActIterT i) : BaseLoop(s,i,"") {}
+    virtual ~IfLoop() {}
+    bool valid() override {return true;}
+    bool next() override {return false;}
+};// IfLoop class
+
+// ==============================================================================================================
+
+struct Parser {
+    ActListT            available, actions;
+    ActIterT            iter;// iterator pointing to the current actions being processed
+    std::unordered_map<std::string, ActIterT> hashMap;
+    std::list<std::shared_ptr<BaseLoop>> loops;
+    std::vector<Option> defaults;
+    int                 verbose;
+    mutable size_t      counter;// loop counter used to validate matching "-for/each" and "-end" actions
+    mutable Processor   processor;// responsible for storing local variables and executing string expressions
+
```

Comment:
As a general comment, the member variables are usually included at the end of the class definition.